src/platform/Infineon/crypto/trustm/CHIPCryptoPALHsm_P256_trustm.cpp

```diff
@@ -0,0 +1,601 @@
+        mInitialized = true;
+    }
+    error = CHIP_NO_ERROR;
+    return error;
+#else    
+    uint8_t pubkey[128]               = {
+        0,
+    };
+    uint16_t pubKeyLen = sizeof(pubkey);
+    optiga_lib_status_t return_status = OPTIGA_LIB_BUSY;
+    P256PublicKey & public_key = const_cast<P256PublicKey &>(Pubkey());
+    optiga_key_usage_t key_usage;
+    uint16_t keyid;
+
+    if (key_target == ECPKeyTarget::ECDH) {   
+        keyid = TRUSTM_ECDH_OID_KEY;
+
+        // Trust M ECC 256 Key Gen
+        ChipLogDetail(Crypto, "Generating NIST256 key in TrustM for ECDH!");
+        key_usage = OPTIGA_KEY_USAGE_KEY_AGREEMENT;
+    } else {
+        // Add the logic to use different keyid
+        keyid = TRUSTM_NODE_OID_KEY_START;
+        // Trust M ECC 256 Key Gen
+        ChipLogDetail(Crypto, "Generating NIST256 key in TrustM !");
+        key_usage = (optiga_key_usage_t)(OPTIGA_KEY_USAGE_SIGN | OPTIGA_KEY_USAGE_AUTHENTICATION);
+    }
+    // Trust M init
+    trustm_Open();
+    return_status = trustm_ecc_keygen(keyid, key_usage, OPTIGA_ECC_CURVE_NIST_P_256, pubkey, &pubKeyLen);
+
+    // Add signature length
+    VerifyOrExit(return_status == OPTIGA_LIB_SUCCESS, error = CHIP_ERROR_INTERNAL);
+
+    /* Set the public key */
+    VerifyOrReturnError((size_t) pubKeyLen > NIST256_HEADER_OFFSET, CHIP_ERROR_INTERNAL);
+    VerifyOrReturnError(((size_t) pubKeyLen - NIST256_HEADER_OFFSET) <= kP256_PublicKey_Length, CHIP_ERROR_INTERNAL);
+    memcpy((void *) Uint8::to_const_uchar(public_key), pubkey + NIST256_HEADER_OFFSET, pubKeyLen - NIST256_HEADER_OFFSET);
+
+    memcpy(&mKeypair.mBytes[0], trustm_magic_no, sizeof(trustm_magic_no));
+    mKeypair.mBytes[CRYPTO_KEYPAIR_KEYID_OFFSET] = (keyid >> (0 * 8)) & 0xFF;
+    mKeypair.mBytes[CRYPTO_KEYPAIR_KEYID_OFFSET + 1] = (keyid >> (1 * 8)) & 0xFF;
+
+    mInitialized = true;
+    error = CHIP_NO_ERROR;
+
+exit:
+    if (error != CHIP_NO_ERROR)
+    {
+        trustm_close();
+    }
+    return error;
+#endif
+}
+
+CHIP_ERROR P256Keypair::ECDSA_sign_msg(const uint8_t * msg, size_t msg_length, P256ECDSASignature & out_signature) const
+{
+#if !ENABLE_TRUSTM_GENERATE_EC_KEY
+    return ECDSA_sign_msg_H(&mKeypair, msg, msg_length, out_signature);
+#else
+    CHIP_ERROR error = CHIP_ERROR_INTERNAL;
+    optiga_lib_status_t return_status = OPTIGA_LIB_BUSY;
+
+    uint8_t signature_trustm[kMax_ECDSA_Signature_Length_Der] = { 0 };
+    uint16_t signature_trustm_len                             = (uint16_t) kMax_ECDSA_Signature_Length_Der;
+    uint8_t digest[32];
+    uint8_t digest_length = sizeof(digest);
+
+    memset(&digest[0], 0, sizeof(digest));
+    MutableByteSpan out_raw_sig_span(out_signature.Bytes(), out_signature.Capacity());
+
+    VerifyOrReturnError(msg != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(msg_length > 0, CHIP_ERROR_INVALID_ARGUMENT);
+    ChipLogDetail(Crypto, "ECDSA_sign_msg: Using TrustM for ecdsa sign!");
+    // Trust M Init
+    trustm_Open();
```

Comment:
it would be better to use RAII around open/close and then you would not combine return macros with gotoexit macros

Reply:
Thanks for your advice. We will improve this in upcoming pull request.